mllib/src/main/scala/org/apache/spark/ml/image/ImageSchema.scala

```diff
@@ -0,0 +1,252 @@
+      } else if (hasAlpha) {
+        (4, ocvTypes("CV_8UC4"))
+      } else {
+        (3, ocvTypes("CV_8UC3"))
+      }
+
+      val imageSize = height * width * nChannels
+      assert(imageSize < 1e9, "image is too large")
+      val decoded = Array.ofDim[Byte](imageSize)
+
+      // Grayscale images in Java require special handling to get the correct intensity
+      if (isGray) {
+        var offset = 0
+        val raster = img.getRaster
+        for (h <- 0 until height) {
+          for (w <- 0 until width) {
+            decoded(offset) = raster.getSample(w, h, 0).toByte
+            offset += 1
+          }
+        }
+      } else {
+        var offset = 0
+        for (h <- 0 until height) {
+          for (w <- 0 until width) {
+            val color = new Color(img.getRGB(w, h))
+
+            decoded(offset) = color.getBlue.toByte
+            decoded(offset + 1) = color.getGreen.toByte
+            decoded(offset + 2) = color.getRed.toByte
+            if (nChannels == 4) {
+              decoded(offset + 3) = color.getAlpha.toByte
+            }
+            offset += nChannels
+          }
+        }
+      }
+
+      // the internal "Row" is needed, because the image is a single DataFrame column
+      Some(Row(Row(origin, height, width, nChannels, mode, decoded)))
+    }
+  }
+
+  /**
+   * :: Experimental ::
+   * Read the directory of images from the local or remote source
+   *
+   * @param path Path to the image directory
+   * @param sparkSession Spark Session
+   * @param recursive Recursive path search flag
+   * @param numPartitions Number of the DataFrame partitions
+   * @param dropImageFailures Drop the files that are not valid images from the result
+   * @param sampleRatio Fraction of the files loaded
+   * @return DataFrame with a single column "image" of images;
+   *         see ImageSchema for the details
+   */
+  def readImages(
+      path: String,
+      sparkSession: SparkSession = null,
+      recursive: Boolean = false,
+      numPartitions: Int = 0,
+      dropImageFailures: Boolean = false,
+      sampleRatio: Double = 1.0): DataFrame = {
+    require(sampleRatio <= 1.0 && sampleRatio >= 0, "sampleRatio should be between 0 and 1")
+
+    val session = if (sparkSession != null) sparkSession else SparkSession.builder().getOrCreate
+    val partitions =
+      if (numPartitions > 0) {
+        numPartitions
+      } else {
+        session.sparkContext.defaultParallelism
+      }
+
+    val oldRecursiveFlag = RecursiveFlag.setRecursiveFlag(Some(recursive.toString), session)
+    val sampleImages = sampleRatio < 1
+    val oldPathFilter: Option[Class[_]] =
+      if (sampleImages) {
+        SamplePathFilter.setPathFilter(Some(classOf[SamplePathFilter]), sampleRatio, session)
+      } else {
+        None
+      }
+
+    try {
+      val streams = session.sparkContext.binaryFiles(path, partitions)
+        .repartition(partitions)
+
+      val convert = (stream: (String, PortableDataStream)) =>
```

Comment:
How about ... :

```scala
val images = if (dropImageFailures) {
  streams.flatMap { case (origin, bytes) => 
    decode(origin, bytes.toArray())
  }
} else {
  streams.map { case (origin, bytes) =>
    // Sets the value for the invalid image if the files that are not valid.
    decode(origin, bytes.toArray()).getOrElse(invalidImageRow(origin)) 
  }
}
```

Reply:
@HyukjinKwon please see comment from @srowen for this code:
"I think you can write this to avoid repeating decode(origin, stream.toArray)"
to resolve I still kept convert method but added the "case (origin, bytes)" to make the code look cleaner